pkg/controller/signer/signer-controller.go

```diff
@@ -0,0 +1,221 @@
+//
+// All requests will be signed using a CA, that is currently generated by
+// the OperatorPKI, and the signed certificate will be returned in the status.
+//
+// This allows clients to get a signed certificate while maintaining
+// private key confidentiality.
+type ReconcileCSR struct {
+	// This client, initialized using mgr.Client() above, is a split client
+	// that reads objects from the cache and writes to the apiserver
+	client client.Client
+	scheme *runtime.Scheme
+	status *statusmanager.StatusManager
+
+	// Note: We need a Clientset as the controller-runtime client does not
+	// support non-CRUD subresources (see
+	// https://github.com/kubernetes-sigs/controller-runtime/issues/452)
+	// This may risk invalidating the cache but in our case, this is not a
+	// problem as we only use this to update the approval status of the csr.
+	clientset *kubernetes.Clientset
+}
+
+// Reconcile CSR
+func (r *ReconcileCSR) Reconcile(request reconcile.Request) (reconcile.Result, error) {
+	csr := &csrv1.CertificateSigningRequest{}
+	err := r.client.Get(context.TODO(), request.NamespacedName, csr)
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			// Request object not found, could have been deleted after reconcile request.
+			// Return and don't requeue as the CSR has been deleted.
+			return reconcile.Result{}, nil
+		}
+		// Error reading the object - requeue the request.
+		log.Println(err)
+
+		return reconcile.Result{}, err
+	}
+
+	// Only handle CSRs for this signer
+	if csr.Spec.SignerName != signerName {
+		// Don't handle a CSR for another signerName. We don't need to log this as
+		// we will pollute the logs. We also don't need to requeue it.
+		return reconcile.Result{}, nil
+	}
+
+	if isCertificateRequestApproved(csr) {
+		// We will not recertify an approved certificate
+		return reconcile.Result{}, nil
+	} else {
+		// We will make the assumption that anyone with permission to issue a
+		// certificate signing request to this signer is automatically approved. This
+		// is somewhat protected by permissions on the CSR resource.
+		// TODO: We may need a more robust way to do this later
+		csr.Status.Conditions = append(csr.Status.Conditions, csrv1.CertificateSigningRequestCondition{
+			Type:    csrv1.CertificateApproved,
+			Status:  "True",
+			Reason:  "AutoApproved",
+			Message: "Automatically approved by " + signerName})
+		// Update status to "Approved"
+		csr, err = r.clientset.CertificatesV1().CertificateSigningRequests().UpdateApproval(context.TODO(), request.Name, csr, metav1.UpdateOptions{})
+		if err != nil {
+			log.Printf("Could not update CSR status: %v", err)
+			return reconcile.Result{}, nil
+		}
+	}
+
+	if len(csr.Status.Certificate) != 0 {
+		// Request already has a certificate. Nothing to do as we will, currently,
+		// not re-certify so don't requeue.
+		log.Printf("Cannot re-sign %s for %s. Please delete and re-create resource", request.Name, signerName)
```

Comment:
Note that this will be called every time CNO restarts, so you might want a different log line. Something like
"Certificate %s already signed; skipping".

Reply:
Done